## HMAS-Allelecalling Pipeline

* [Introduction](#introduction)
* [Preparation](#preparation)
* [Execution](#execution)

### Introduction
HMAS-Allelecalling pipeline is designed to identify alleles in metagenomics samples using highly-multiplexed amplicon sequencing data. This pipeline does reference based BLAST search for allele calling. It acts as a denoising step for being able to identify different variants of Salmonella with higher accuracy in the downstream analysis.


![Screenshot (27)](https://user-images.githubusercontent.com/93733968/228165014-f6c069b0-66eb-4478-832d-149f284869f7.png)

### Preparation
Allele calling pipeline requires the following pre-requisites:

* **Predicted primers database**

A set of insilico designed primer sequences is selected as a training set. This set of sequences in fasta format is run through the following command to create a database for sorting *Input sequences* into *Known Alleles* and *Unknown Alleles*.

      makeblastdb -in BLAST1fasta -title BLAST1_DB -dbtype nucl -out /path to save/BLAST1_DB


* **Centroid reference fasta file**

Centroid reference fasta file is a set of sequences generated by following the centroid reference generation protocol explained [here](https://github.com/aminaDBM/HMAS-Allelecalling-Pipeline/tree/main/Centroid_Reference)

![Screenshot (30)](https://user-images.githubusercontent.com/93733968/228158232-1e36dd16-1b11-450a-ba44-62388db17930.png)

* **Centroid Reference database**

Centroid reference fasta file is utilized to construct the *Centroid_Reference database* . The set of centroid reference sequences in fasta format is run through the following command:

            makeblastdb -in BLAST2fasta -title BLAST2_DB -dbtype nucl -out /path to save/BLAST2_DB

Centroid reference database helps sort the *Unknown alleles* into *Novel alleles* and *Not found alleles*.


### Pipeline Workflow
Allele calling pipeline consists of four integral steps defined as follows:

![Screenshot (32)](https://user-images.githubusercontent.com/93733968/228159991-b35b6d77-dac5-434d-b569-1eac641194f2.png)

#### Step1:

Run BLAST on input "fasta" files.
* INPUT: fasta files, InsilicoPrimers database
* OUTPUT: tsv files

#### Step2:

Run Rscript "Known.alleles-pipelinepart1.R" to apply filters on the first BLAST results and identify "Known Alleles".
* INPUT: fasta files, tsv files
* OUTPUT: KnownAlleles.fasta, KnownAlleleStatistics.csv, KnownAlleles.consolidatedStats.csv, UnknownAlleles.fasta

#### Step3:

Run BLAST on "Unknown Alleles" fasta files
* INPUT: UnknownAlleles fasta files, Centroid Reference database
* OUTPUT: UnknownAlleles.tsv files

#### Step4:

Run Rscript "Novel.alleles-pipelinepart2.R" to apply filters on second BLAST results and identify "Novel Alleles".
* INPUT: UnknownAlleles fasta files, UnknownAlleles.tsv files
* OUTPUT: NovelAlleles.fasta, NovelAlleleStatistics.csv, NovelAlleles.consolidatedStats.csv, NotfoundAlleles.fasta

### Execution

The pipeline comprises of two R scripts and a bash script kept in the same path. The bash script requires the user to specify paths to input data, Insilico primers database and centroid reference database. 
The pipeline can be initiated using the following command:

      bash Allelecalling.sh
  
