## HMAS-Allelecalling Pipeline

### Preparation
Allelecalling pipeline requires the following pre-requisites:
* **Insilico primers database**
*A set of insilico designed primer sequences is selected as a training set. This set of sequences in fasta format is run through "makeDatabase.sh" to create a database for sorting input sequences into "Known Alleles" and "Unknown Alleles".
* **Centroid reference fasta file**
Centroid reference fasta file is a set of sequences generated by following the centroid reference generation protocol explained [here](https://github.com/aminaDBM/HMAS-Allelecalling-Pipeline/tree/main/Centroid_Reference)
* Centroid Reference database
![Screenshot (27)](https://user-images.githubusercontent.com/93733968/228119977-c15fa1ab-7025-4b26-b957-daeef87e0622.png)


### Pipeline Workflow
Allelecalling pipeline consists of four integral steps defined as follows:

#### Step1:

Run BLAST on input "fasta" files.
* INPUT: fasta files, InsilicoPrimers database
* OUTPUT: tsv files

#### Step2:

Run Rscript "Known.alleles-pipelinepart1.R" to apply filters on the first BLAST results and identify "Known Alleles".
* INPUT: fasta files, tsv files
* OUTPUT: KnownAlleles.fasta, KnownAlleleStatistics.csv, KnownAlleles.consolidatedStats.csv, UnknownAlleles.fasta

#### Step3:

Run BLAST on "Unknown Alleles" fasta files
* INPUT: UnknownAlleles fasta files, Centroid Reference database
* OUTPUT: UnknownAlleles.tsv files

#### Step4:

Run Rscript "Novel.alleles-pipelinepart2.R" to apply filters on second BLAST results and identify "Novel Alleles".
* INPUT: UnknownAlleles fasta files, UnknownAlleles.tsv files
* OUTPUT: NovelAlleles.fasta, NovelAlleleStatistics.csv, NovelAlleles.consolidatedStats.csv, NotfoundAlleles.fasta

### Execution

The pipeline comprises of two R scripts and a bash script kept in the same path. The bash script requires the user to specify paths to input data, Insilico primers database and centroid reference database. 
The pipeline can be initiated using the following command:

      bash Allelecalling.sh
  
